File: src/main/kotlin/com/huawei/cangjie/types/checker/CangJieTypeChecker.java
package com.huawei.cangjie.types.checker;

import com.huawei.cangjie.types.CangJieType;
import com.huawei.cangjie.types.TypeConstructor;
import org.jetbrains.annotations.NotNull;


public interface CangJieTypeChecker {

    interface TypeConstructorEquality {
        boolean equals(@NotNull TypeConstructor a, @NotNull TypeConstructor b);
    }

    CangJieTypeChecker DEFAULT = NewCangJieTypeChecker.Companion.getDefault();

    /**
     * 比较两个类型，但是不比较泛型
     */
    boolean equalsIgnoringGenerics(@NotNull CangJieType a, @NotNull CangJieType b);

    boolean isSubtypeOf(@NotNull CangJieType subtype, @NotNull CangJieType supertype);
    boolean equalTypes(@NotNull CangJieType a, @NotNull CangJieType b);
}

File: src/main/kotlin/com/huawei/cangjie/types/checker/CangJieTypeChecker.kt
package com.huawei.cangjie.types.checker

import com.huawei.cangjie.types.CangJieType
import com.huawei.cangjie.types.TypeConstructor

interface CangJieTypeChecker {
    interface TypeConstructorEquality {
        fun equals(
            a: TypeConstructor,
            b: TypeConstructor
        ): Boolean
    }

    fun isSubtypeOf(
        subtype: CangJieType,
        supertype: CangJieType
    ): Boolean

    fun equalTypes(a: CangJieType, b: CangJieType): Boolean

    companion object {
        @JvmField
        val DEFAULT: CangJieTypeChecker = NewCangJieTypeChecker.Default
    }
}
